core-to-plc/src/Language/Plutus/CoreToPLC.hs

```diff
@@ -0,0 +1,717 @@
+import qualified Data.Map                                 as Map
+import qualified Data.Text                                as T
+import qualified Data.Text.Encoding                       as TE
+
+import           Data.List                                (elemIndex)
+
+{- Note [System FC and System FW]
+Haskell uses system FC, which includes type equalities and coercions.
+
+PLC does *not* have coercions in particular. However, PLC also does not have a nominal
+type system - everything is constructed via operators on base types, so we have no
+need for coercions to convert between newtypes and datatypes.
+
+So we mostly ignore coercions. The one place that I know of where the mismatch hurts
+us is that GHC uses the `Any` type (coercible to and from anything) for unconstrained
+function variables, e.g. in polymorphic lambdas. This is annoying for us, since we
+really want the version with explicit type abstraction. I don't currently have a fix
+for this.
+-}
+
+{- Note [Iterated abstraction and application]
+PLC doesn't expose iterated abstraction and application.
+
+We typically build these with a fold.
+- Iterated application uses a *left* fold, since we want to apply the first variable
+first.
+- Iterated abstraction uses a *right* fold, since we want to bind the first
+variable *last* (so it is on the outside, so will be first when applying).
+-}
+
+-- Useful synonyms and functions
+
+type PCExpr = PC.Term PC.TyName PC.Name ()
+type PCType = PC.Type PC.TyName ()
+
+type PrimTerms = Map.Map GHC.Name (Quote PCExpr)
+type PrimTypes = Map.Map GHC.Name (Quote PCType)
+
+type ConvertingState = (GHC.DynFlags, PrimTerms, PrimTypes, ScopeStack)
+-- See Note [Scopes]
+type Converting m = (Monad m, MonadError (Error ()) m, MonadQuote m, MonadReader ConvertingState m)
+
+strToBs :: String -> BSL.ByteString
+strToBs = BSL.fromStrict . TE.encodeUtf8 . T.pack
+
+bsToStr :: BSL.ByteString -> String
+bsToStr = T.unpack . TE.decodeUtf8 . BSL.toStrict
+
+sdToTxt :: (MonadReader ConvertingState m) => GHC.SDoc -> m T.Text
+sdToTxt sd = do
+  (flags, _, _, _) <- ask
+  pure $ T.pack $ GHC.showSDoc flags sd
+
+conversionFail :: (MonadError (Error ()) m, MonadReader ConvertingState m) => GHC.SDoc -> m a
+conversionFail = (throwError . ConversionError) <=< sdToTxt
+
+unsupported :: (MonadError (Error ()) m, MonadReader ConvertingState m) => GHC.SDoc -> m a
+unsupported = (throwError . UnsupportedError) <=< sdToTxt
+
+freeVariable :: (MonadError (Error ()) m, MonadReader ConvertingState m) => GHC.SDoc -> m a
+freeVariable = (throwError . FreeVariableError) <=< sdToTxt
+
+-- Names and scopes
+
+{- Note [Scopes]
+We need a notion of scope, because we have to make sure that if we convert a GHC
+Var into a variable, then we always convert it into the same variable, while also making
+sure that if we encounter multiple things with the same name we produce fresh variables
+appropriately.
+
+So we have the usual mechanism of carrying around a stack of scopes.
+-}
```

Comment:
Do you really need that machinery? `GHC.Name` already embeds a unique — see https://downloads.haskell.org/~ghc/latest/docs/html/libraries/ghc-8.4.3/Name.html

If we generate a PLC name including the unique, all the scoping *for names defined in Core* ought to already be taken care of, or am I overlooking anything here? (Then, you would only need fresh names for new names introduced during code gen.)

Moreover, instead of a stack of scopes, a simple finite map will do.

Reply:
We might not need a stack. But it's *definitely* safe, whereas I'd rather not reason about whether the global one is safe.

GHC names do have a unique attached, and originally I just translated that into our uniques. However, I think it's much more clearly correct to just generate a fresh name corresponding to each GHC name - in practice this uses the GHC unique so the resulting name is pretty unique anyway, but I'm sure it's correct.

Reply:
I'd suggest to instead to [tidy](https://downloads.haskell.org/~ghc/latest/docs/html/libraries/ghc-8.4.3/CoreTidy.html) the core expression. Then we know everything has been renamed, and we don't have any name clashes. (It seems better to me to let GHC do the work where we can.)